libgalois/include/katana/analytics/random_walks/random_walks.h

```diff
@@ -0,0 +1,106 @@
+
+  // Don't allow people to directly construct these, so as to have only one
+  // consistent way to configure.
+private:
+  Algorithm algorithm_;
+  uint32_t walk_length_;
+  uint32_t number_of_walks_;
+  double backward_probability_;
+  double forward_probability_;
+  // Only need for edge2vec
+  uint32_t max_iterations_;
+  // TODO(gill) Find number of edges automatically
+  uint32_t number_of_edge_types_;
+
+  RandomWalksPlan(
+      Architecture architecture, Algorithm algorithm, uint32_t walk_length,
+      uint32_t number_of_walks, double backward_probability,
+      double forward_probability, uint32_t max_iterations,
+      uint32_t number_of_edge_types)
+      : Plan(architecture),
+        algorithm_(algorithm),
+        walk_length_(walk_length),
+        number_of_walks_(number_of_walks),
+        backward_probability_(backward_probability),
+        forward_probability_(forward_probability),
+        max_iterations_(max_iterations),
+        number_of_edge_types_(number_of_edge_types) {}
+
+public:
+  // kChunkSize is a fixed const int (default value: 1)
+  static const int kChunkSize;
+
+  RandomWalksPlan() : RandomWalksPlan{kCPU, kNode2Vec, 1, 1, 1.0, 1.0, 10, 1} {}
+
+  Algorithm algorithm() const { return algorithm_; }
+  uint32_t walk_length() const { return walk_length_; }
+  uint32_t number_of_walks() const { return number_of_walks_; }
+  double backward_probability() const { return backward_probability_; }
+  double forward_probability() const { return forward_probability_; }
+  uint32_t max_iterations() const { return max_iterations_; }
+  uint32_t number_of_edge_types() const { return number_of_edge_types_; }
+
+  /// Node2Vec algorithm to generate random walks on the graph
+  static RandomWalksPlan Node2Vec(
+      uint32_t walk_length, uint32_t number_of_walks,
+      double backward_probability, double forward_probability) {
+    return {
+        kCPU,
+        kNode2Vec,
+        walk_length,
+        number_of_walks,
+        backward_probability,
+        forward_probability,
+        0,
+        1};
+  }
+
+  /// Edge2Vec algorithm to generate random walks on the graph.
+  /// Takes the heterogeneity of the edges into account
+  static RandomWalksPlan Edge2Vec(
+      uint32_t walk_length, uint32_t number_of_walks,
+      double backward_probability, double forward_probability,
+      uint32_t max_iterations, uint32_t number_of_edge_types) {
+    return {
+        kCPU,
+        kNode2Vec,
+        walk_length,
+        number_of_walks,
+        backward_probability,
+        forward_probability,
+        max_iterations,
+        number_of_edge_types};
+  }
+};
+
+/// Compute the random-walks for pg. The pg is expected to be
+/// symmetric.
+/// The parameters can be specified, but have reasonable defaults. Not all parameters
+/// are used by the algorithms.
+/// The generated random-walks generated are return in Katana::InsertBag.
+KATANA_EXPORT Result<katana::InsertBag<katana::gstl::Vector<uint32_t>>>
```

Comment:
I'll review when you actually ask me too. But I thought I would warn you: There is Python support for `InsertBag` but only for primitive types and simple structs of primitive types. This type would need to be fully converted before it can be returned. If we can simplify it to be a bit more C++ and less Katana (maybe use `std::vector<std::vector<...>>`) that would make Python wrapping easier.

Reply:
In that  case, we would need to create a new `std::vector` and move the elements of the `InsertBag` to it. We still need `InsertBag` for our random-walk generation.

Reply:
@arthurp Thank you for pointing that out. Let us do a quick performance evaluation using std::vector.

Reply:
I could see an argument that using `vector` is also better for C++ because it is a well known type. Internally you would still use InsertBag and just convert before you return. Is that what you are going to benchmark?

Reply:
I thought what he meant was using `std::vector` instead of `gstl::Vector` during the random-walk computation. I will do that.

Reply:
Yes, I meant just replacing gstl::Vector altogether but I see @arthurp point of just copying over results before returning. We didn't notice any significant performance degradation with using std::vector though.